include/oc_push.h

```diff
@@ -0,0 +1,232 @@
+	/* push proxy */
+	oc_string_t pushtarget_di;    ///< device id of target (e.g. ocf://17087f8c-13e3-4849-4258-65af2a47df63)
+	oc_endpoint_t pushtarget_ep;  ///< endpoint of pushtarget (e.g. coaps://[fe80::b1d6]:1122)
+	oc_string_t targetpath;       ///< path in pushtarget (e.g. /pushedLightSwitch)
+	oc_string_t pushqif;          ///< oic.r.pushproxy:pushqif (deprecated)
+	oc_string_array_t sourcert;   ///< oic.r.pushproxy:sourcert
+	oc_string_t state;            ///< oic.r.pushproxy:state
+	void *user_data;              ///< used to point updated pushable Resource
+} oc_ns_t;
+
+
+/**
+ * @brief structure for member of "oic.r.pushreceiver:receivers" object array
+ */
+typedef struct oc_recv
+{
+	struct oc_recv *next;
+	oc_string_t receiveruri;  ///< oic.r.pushreceiver:receivers:receiveruri
+	oc_string_array_t rts;    ///< oic.r.pushreceiver:receivers:rts
+} oc_recv_t;
+
+
+/**
+ * @brief structure for handling for Push Receiver Resource
+ */
+typedef struct oc_recvs
+{
+	struct oc_recvs *next;
+	oc_resource_t *resource;    ///< used to point ["oic.r.pushreceiver"] Resource managed by iotivity-lite
+	OC_LIST_STRUCT(receivers);  ///< oic.r.pushreceiver:receivers object array
+} oc_recvs_t;
+
+
+
+/**
+ * @brief object used to store Resource pushed to "oic.r.pshreceiver:receivers[i].receiveruri"
+ */
+typedef struct oc_pushd_rsc_rep
+{
+	struct oc_pushd_rsc_rep *next;
+	oc_resource_t *resource;        ///< used to point any pushed Resource managed by iotivity-lite
+	oc_rep_t *rep;                  ///< payload of pushed Resource
+} oc_pushd_rsc_rep_t;
+
+
+OC_PROCESS_NAME(oc_push_process);
+
+extern char *pp_state_strs[];
+extern char *cli_status_strs[];
+
+/**
+ * @brief if this callback function is provided by user, it will called whenever new push notification arrives...
+ */
+extern void (*oc_push_arrived)(oc_pushd_rsc_rep_t *);
+
+
+#define cli_statusstr(i) (cli_status_strs[(i)])
+#define pp_statestr(i) (pp_state_strs[(i)])
+
+/**
+ * @brief update Push Proxy state from state to new_state
+ *
+ * @param state		oc_string_t
+ * @param new_state	char *
+ */
+#define pp_update_state(state, new_state) \
+{ \
+	oc_free_string(&(state)); \
+	oc_new_string(&(state), (new_state), strlen((new_state))); \
+}
+
+/**
+ * @brief initialize oc_string_t object
+ */
+#define oc_init_string(str) \
+{ \
+	(str).size = 0; \
+	(str).ptr = NULL; \
+	(str).next = NULL; \
+}
+
+/**
+ * @brief initialize OC_LIST for push notification
+ */
+void oc_push_list_init();
+
+/**
+ * @brief clean up push related data structure
+ */
+void oc_push_free();
+
+/**
+ * @brief initialize Push Configuration Resource
+ */
+void oc_create_pushconf_resource(size_t device_index);
+
+/**
+ * @brief initiate Push Receiver Resource
+ */
+void oc_create_pushreceiver_resource(size_t device_index);
+
+/**
+ * @brief try to find `receiver` object which has `uri` as its `uri` Property
+ */
+oc_recv_t * _find_recv_obj_by_uri(oc_recvs_t *recvs_instance, const char *uri, int uri_len);
```

Comment:
This doesn't seem like a public function, so we shouldn't expose the declaration in a public header. It seems to be used in just `oc_push.c` so just add the `static` keyword to the definition in the `oc_push.c` file.

Reply:
fixed